### ISSUE TITLE
shadow/matte material - no render, only shadow, reflection etc.

### ISSUE THREAD

**Issue by Gibbz** [godotengine/godot#9552]
So this trick has been around for many years in software such as blender and 3ds max. They have a material that receives shadows, or reflections and effects. As the user chooses.

It would be really handy to have a material like this! Im working on something where I need an invisible ground plane. But I want reflections and shadows on the invisible ground.....

Is there anything like this in gd3?

**Comment by rredesigns** [comment]
How could you have reflections on something that is not visible? Doesn't make sense to me. If it is some kind of mirror then it couldn't be invisible anyway.

You might get away with that using just a reflective plane.

**Comment by supagu** [comment]
so for example

Say I have an image of a desk. I want to overlay a 3d mug on that desk. But I want to have a plane for the shadow and reflection to appear to make it look like they are falling on the desk. While the desk is in 2d space, i need a shadow and reflection to appear without blocking the 2d image of the desk.

**Comment by rredesigns** [comment]
That's called a shadow catcher. :)

**Comment by reduz** [comment]
you can do this fine by sampling the screen texture currently

**Comment by pgregory** [comment]
Would really appreciate some guidance on how to do this. I don't understand the comment "sampling the screen texture", any step by step tutorial or at least pointers would be great. I've got a 2D game that would really shine with some 3D in it, but I need to have the 3D cast a shadow onto the 2D background.

**Comment by siamak-s** [comment]
Any idea about Shadow Catcher on Godot? or any tutorial on @reduz idea's?
It seems this issue doesn't take any decision yet, no bug label, no milestone, no idea ...

**Comment by siamak-s** [comment]
I found that!
Simple shader something like below can do Shadow Capturing:
`shader_type spatial;

void fragment() {
	//ALBEDO = vec3(1.0, 0, 0);
	ALBEDO = texture(SCREEN_TEXTURE, SCREEN_UV).rgb;
	//ALPHA = 0.0;
}`

**Comment by akien-mga** [comment]
Feature and improvement proposals for the Godot Engine are now being discussed and reviewed in a dedicated [**Godot Improvement Proposals (GIP)**](https://github.com/godotengine/godot-proposals) ([godotengine/godot-proposals](https://github.com/godotengine/godot-proposals)) issue tracker. The GIP tracker has a detailed issue template designed so that proposals include all the relevant information to start a productive discussion and help the community assess the validity of the proposal for the engine.

The main ([godotengine/godot](https://github.com/godotengine/godot)) tracker is now solely dedicated to bug reports and Pull Requests, enabling contributors to have a better focus on bug fixing work. Therefore, we are now closing all older feature proposals on the main issue tracker.

If you are interested in this feature proposal, please [**open a new proposal on the GIP tracker**](https://github.com/godotengine/godot-proposals/issues/new/choose) following the given issue template (after checking that it doesn't exist already). Be sure to reference this closed issue if it includes any relevant discussion (which you are also encouraged to summarize in the new proposal). Thanks in advance!

**Comment by clayjohn** [comment]
Note, that this suggestion is already in the engine as siamak-s pointed out.

**Comment by sygi** [comment]
In case someone stumbled here [as I did](https://godotengine.org/qa/123307/shadow-catcher-doesnt-catch-shadows) with a problem that the siamak-s solution makes the non-shadow part not fully transparent (it seems to catch the reflections from the lights, which I didn't want), I made this shadow catcher that modifies ALPHA instead:
```
shader_type spatial;

void fragment(){
    ALBEDO=0.02 * vec3(1.);
}

void light(){
    // There needs to be something touching DIFFUSE_LIGHT for the shader to be calculated at all:
    DIFFUSE_LIGHT = DIFFUSE_LIGHT;
    ALPHA = clamp(1. - length(ATTENUATION), 0, 1.);
}
```

**Comment by Calinou** [comment]
This is available since Godot 3.1 with the **Shadow To Opacity** flag in SpatialMaterial.

**Comment by sygi** [comment]
Thanks! I was looking there for inspiration, but I wasn't able to make the color of the shadow black as opposed to the color of the object, but now I see I could have just change the ALBEDO in the settings (facepalm emoji here).

Do you have an idea whether it's possible to make the material actually catch the shadow, though? Both shadow to opacity and what I propose above and what siamak-s proposed cause the shadow to appear on the material, but don't stop it from being rendered in the underlying background, again.
